### PR TITLE
security events feed

### PR DESCRIPTION
**Pull request by jaredh159**
closes https://github.com/gertrude-app/project/issues/253

### CODE REVIEW

File: dash/app/src/components/routes/SecurityEventsFeed.tsx [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+import { Link } from 'react-router-dom';
+import type { SecurityEventsFeed } from '@dash/types';
+import { Key, useQuery } from '../../hooks';
+import Current from '../../environment';
+// import { Key, useQuery } from '../../hooks';
+// import Current from '../../environment';
+// import { ApiErrorMessage, Loading } from '@dash/components';
+
+const SecurityEventsFeed: React.FC = () => (
+  // const getFeed = useQuery(Key.securityEventsFeed, Current.api.securityEventsFeed);
```
Comment (kiahjh) [inline dash/app/src/components/routes/SecurityEventsFeed.tsx:14]: I just hard-coded in the values from the json you sent me for now.

File: dash/types/src/pairql/pairs/SecurityEventsFeed.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+export namespace SecurityEventsFeed {
+  export type Input = void;
+
+  export type Output = Array<
+    | {
+        case: 'child';
+        id: UUID;
+        childId: UUID;
+        deviceId: UUID;
```
Comment (jaredh159) [inline dash/types/src/pairql/pairs/SecurityEventsFeed.ts:11]: add two fields here temporarily for typescript until i redo the codegen:

```
   childName: string;
   deviceName: string;
```

File: dash/app/src/components/routes/SecurityEventsFeed.tsx [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+  const [where, setWhere] = useState<{
+    city: string;
+    region: string;
+    country_code: string;
+  } | null>(null);
+
+  const childId = event.case === `child` ? event.childId : ``;
+  const deviceId = event.case === `child` ? event.deviceId : ``;
+  const childQuery = useQuery(Key.user(childId), () => Current.api.getUser(childId));
+  const deviceQuery = useQuery(Key.device(deviceId), () =>
```
[the remark this answers is not in the document]
Comment (jaredh159) [reply dash/app/src/components/routes/SecurityEventsFeed.tsx:63]: so, delete both these queries, make the other changes commented below, and use the json attached to this

[new.json](https://github.com/user-attachments/files/16591120/new.json)

File: dash/components/src/Chrome/SidebarNav.tsx [hunk 4]
```diff
@@ -79,6 +80,15 @@ const SidebarNav: React.FC<Props> = ({
...
         </SidebarOption>
+        <SidebarOption
+          Icon={ShieldCheckIcon}
+          collapsed={collapsed}
+          to="/security-events"
+          onClick={onInternalLinkClick}
+          isSelected={urlPath.startsWith(`/security-events`)}
+        >
+          Security
+        </SidebarOption>
```
Comment (jaredh159) [inline dash/components/src/Chrome/SidebarNav.tsx:91]: this looks nice, thanks. however, can you comment it out for now? i'd like to use the feature myself for a week or two before exposing the link. and i'm not 100% sure it should be a top-level sidebar link yet either. but if it does become one, this looks great, so just comment it out for now, i think. 👍

File: dash/app/src/components/routes/SecurityEventsFeed.tsx [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+const SecurityEvent: React.FC<{ event: SecurityEventsFeed.Output[number] }> = ({
+  event,
+}) => {
+  const whenItHappened = new Date(event.createdAt);
+  const [explanationExpanded, setExplanationExpanded] = useState(false);
+  const [where, setWhere] = useState<{
+    city: string;
+    region: string;
+    country_code: string;
+  } | null>(null);
```
[the remark this answers is not in the document]
Comment (jaredh159) [reply dash/app/src/components/routes/SecurityEventsFeed.tsx:58]: just to be more clear, what i'm proposing is that we always show the raw ip address as a link to whatismyipaddress.com, and then if we can get good data from ipapi.co and we're not rate limited, we add the fancy "Wadsworth, OH, USA" stuff in addition.

**Review by jaredh159 — COMMENTED** [review]
it's not letting me "request changes" because i opened the PR originally, but I am requesting changes. :) 

looks good, thanks. yeah, it's a bit hard to design, but i think this is pretty solid, and i can't think of way to improve it at the moment.

let me know if my comments are confusing at all.

**Comment by jaredh159** [comment]
awesome, thanks. looks nice.

i ripped out the hard-coded data and memoized the api lookup (since it's likely there will be tons of dupes), but otherwise left it as is. 👍

**Pull request merged**